pkg/kubelet/kuberuntime/logs/logs_test.go

```diff
@@ -210,6 +212,157 @@ func TestReadLogs(t *testing.T) {
 	}
 }
 
+type syncBuffer struct {
+	b bytes.Buffer
+	sync.Mutex
+}
+
+func (b *syncBuffer) Read(p []byte) (n int, err error) {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.Read(p)
+}
+func (b *syncBuffer) Write(p []byte) (n int, err error) {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.Write(p)
+}
+
+func (b *syncBuffer) String() string {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.String()
+}
+
+func (b *syncBuffer) Reset() {
+	b.Lock()
+	defer b.Unlock()
+	b.b.Reset()
+}
+
+func TestReadRotatedLog(t *testing.T) {
+	tmpDir := t.TempDir()
+
+	file, err := os.CreateTemp(tmpDir, "logfile")
+	if err != nil {
+		t.Errorf("unable to create temp file")
+		t.Fail()
+	}
+
+	fileName := file.Name()
+	podLogOptions := v1.PodLogOptions{
+		Follow: true,
+	}
+
+	containerID := "fake-container-id"
+	fakeRuntimeService := &apitesting.FakeRuntimeService{
+		Containers: map[string]*apitesting.FakeContainer{
+			containerID: {
+				ContainerStatus: runtimeapi.ContainerStatus{
+					State: runtimeapi.ContainerState_CONTAINER_RUNNING,
+				},
+			},
+		},
+	}
+
+	opts := NewLogOptions(&podLogOptions, time.Now())
+	stdoutBuf := &syncBuffer{}
+	stderrBuf := &syncBuffer{}
+
+	// log in stdout
+	expectedStdout := "line1\nline2\nline3\nline4\nline5\nline7\n"
+	// log in stderr
+	expectedStderr := "line6\nline8\n"
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dir := filepath.Dir(file.Name())
+	baseName := filepath.Base(file.Name())
+
+	stopped := make(chan struct{})
+	go func() {
+		timeout := time.NewTimer(30 * time.Second)
+		ticker := time.NewTicker(10 * time.Millisecond)
```

Comment:
I don't fully get the idea of using timer and ticker. Can you elaborate?

Would it be simpler to simply write to the file, rotate it and then write to it again in one goroutine and call ReadLogs in another one?